util/coverage/common.py

```diff
@@ -0,0 +1,279 @@
+    instrumentation, gathers and returns their object files which can then be combined
+    into a single library to avoid such problems.
+
+    Args:
+        device_libs: List of device libraries to be instrumented.
+        config: Bazel configuration to use.
+
+    Returns:
+        Instrumented object files of the given device libraries.
+    """
+    log.info(f"libraries to be instrumented: {pformat(device_libs)}")
+    run(BAZEL, "build", f"--config={config}", *device_libs)
+    obj_files = run(
+        BAZEL, "cquery", f"--config={config}", f"({' + '.join(device_libs)})",
+        "--output=starlark",
+        ("--starlark:expr='\\n'.join("
+         "[f.path for f in target.output_groups.compilation_outputs.to_list()]"
+         ")"))
+    log.info(f"object files with coverage data: {pformat(obj_files)}")
+    return obj_files
+
+
+def get_test_log_dirs(test_targets: List[str]) -> List[Path]:
+    """Get log direactories for the given test targets.
+
+    Args:
+        test_targets: Test targets.
+
+    Returns:
+        Log directories of the the given test targets.
+    """
+    [test_log_dir] = run(
+        BAZEL,
+        "info",
+        "bazel-testlogs",
+    )
+    return [Path(test_log_dir) / t[2:].replace(':', '/') for t in test_targets]
+
+
+def generate_report(out_dir: Path, merged_profile: Path, merged_library: Path,
+                    report_title: str, print_text_report: bool) -> None:
+    """Generate a coverage report.
+
+    This function generates a coverage report from the given merged profile and library.
+
+    Args:
+        out_dir: Path where to save the html report.
+        merged_profile: Path of merged profile data.
+        merged_library: Path of the merged library with coverage data.
+        print_text_report: Whether to print the text report.
+    """
+    # Generate html coverage report.
+    run(LLVM_COV, "show", "--show-line-counts", "--show-regions",
+        f"--project-title={report_title}", "--format=html",
+        f"--output-dir=./{out_dir}", "--instr-profile", str(merged_profile),
+        str(merged_library))
+    # Generate text coverage report.
+    with (out_dir / "report.txt").open("w") as f:
+        f.write("\n".join(
+            run(LLVM_COV, "report", "--instr-profile", str(merged_profile),
+                str(merged_library))))
+    if print_text_report:
+        print("\n".join(
+            run(LLVM_COV, "report", "--use-color", "--instr-profile",
+                str(merged_profile), str(merged_library))))
+    print(f"Saved coverage artifacts in {out_dir}")
+
+
+# Query for unit test targets
+TEST_TARGETS_INC = [
+    "//sw/device/silicon_creator/...",
+]
+TEST_TARGETS_EXC = [
+    "//sw/device/silicon_creator/rom_ext/...",
+    "attr(tags, 'dv', //sw/device/silicon_creator/...)",
+    "attr(tags, 'verilator', //sw/device/silicon_creator/...)",
+    "attr(tags, 'broken', //sw/device/silicon_creator/...)",
+]
```

Comment:
Aha, I think we should do the same thing as [BazelQuery.tag_exact()](https://cs.opensource.google/opentitan/opentitan/+/master:ci/scripts/lib/bazel_query.py;l=19;drc=465f8939c4fae38dff384cb0d0a032508862b593) here to restrict to exact matches.  Plain 'dv' will match anything that contains 'dv', e.g. 'advice' (I don't believe this a real tag, but it illustrates my point).

Reply:
For this and the previous comment I just inlined the fixes but we should move BazelQuery to `util` when we get the chance. Thanks for catching these!